From d24290b18729c3f00d16330e4941a91c9263ddd0 Mon Sep 17 00:00:00 2001
Message-Id: <d24290b18729c3f00d16330e4941a91c9263ddd0.1765541409.git.gerbach@fs-net.de>
In-Reply-To: <1136d5984def64a1ed82cb36ba313f2651b7c130.1765541409.git.gerbach@fs-net.de>
References: <1136d5984def64a1ed82cb36ba313f2651b7c130.1765541409.git.gerbach@fs-net.de>
From: Philipp Gerbach <gerbach@fs-net.de>
Date: Tue, 13 May 2025 16:53:22 +0200
Subject: [PATCH 07/11] Add support for the EE0350ET F&S MIPI display

Move display intitalisation to enable function, because the DSI bridge
is not initialized when the display prepare is called.
---
 .../gpu/drm/panel/panel-newvision-nv3051d.c   | 312 ++++++++++++++++--
 1 file changed, 286 insertions(+), 26 deletions(-)

diff --git a/drivers/gpu/drm/panel/panel-newvision-nv3051d.c b/drivers/gpu/drm/panel/panel-newvision-nv3051d.c
index 5d115ecd5..d05835760 100644
--- a/drivers/gpu/drm/panel/panel-newvision-nv3051d.c
+++ b/drivers/gpu/drm/panel/panel-newvision-nv3051d.c
@@ -23,14 +23,6 @@
 #include <drm/drm_modes.h>
 #include <drm/drm_panel.h>
 
-struct nv3051d_panel_info {
-	const struct drm_display_mode *display_modes;
-	unsigned int num_modes;
-	u16 width_mm, height_mm;
-	u32 bus_flags;
-	u32 mode_flags;
-};
-
 struct panel_nv3051d {
 	struct device *dev;
 	struct drm_panel panel;
@@ -39,6 +31,16 @@ struct panel_nv3051d {
 	struct regulator *vdd;
 };
 
+struct nv3051d_panel_info {
+	const struct drm_display_mode *display_modes;
+	unsigned int num_modes;
+	u16 width_mm, height_mm;
+	u32 bus_flags;
+	u32 mode_flags;
+	int (*init_sequence)(struct panel_nv3051d *ctx);
+	unsigned int lanes;
+};
+
 static inline struct panel_nv3051d *panel_to_panelnv3051d(struct drm_panel *panel)
 {
 	return container_of(panel, struct panel_nv3051d, panel);
@@ -243,6 +245,222 @@ static int panel_nv3051d_init_sequence(struct panel_nv3051d *ctx)
 	return 0;
 }
 
+static int panel_nv3051d_ee0350et_init_sequence(struct panel_nv3051d *ctx)
+{
+	struct mipi_dsi_device *dsi = to_mipi_dsi_device(ctx->dev);
+	struct mipi_dsi_multi_context dsi_ctx = {.dsi = dsi};
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x30);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x52);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xE3,0x00);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x40,0x0a);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x03,0x40);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x04,0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x05,0x03);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x20,0x90);//2LANE
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x24,0x0c);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x25,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x26,0x14);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x27,0x14);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x28,0x57);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x29,0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x2A,0xdf);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x38,0x9C);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x39,0xA7);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x3A,0x53);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x44,0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x49,0x3C);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x59,0xfe);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x5c,0x00);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x91,0x57);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x92,0x57);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA0,0x55);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA1,0x50);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA4,0x9C);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA7,0x02);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA8,0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA9,0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xAA,0xFC);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xAB,0x28);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xAC,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xAD,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xAE,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xAF,0x03);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB0,0x08);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB1,0x26);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB2,0x28);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB3,0x28);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB4,0x33);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB5,0x08);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB6,0x26);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB7,0x08);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB8,0x26);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x30);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x52);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x02);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB0,0x0B);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB1,0x16);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB2,0x17);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB3,0x2C);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB4,0x32);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB5,0x3B);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB6,0x29);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB7,0x40);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB8,0x0d);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB9,0x05);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBA,0x12);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBB,0x10);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBC,0x12);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBD,0x15);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBE,0x19);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBF,0x0E);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xC0,0x16);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xC1,0x0A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD0,0x0C);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD1,0x17);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD2,0x14);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD3,0x2E);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD4,0x32);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD5,0x3C);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD6,0x22);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD7,0x3D);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD8,0x0D);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD9,0x07);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xDA,0x13);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xDB,0x13);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xDC,0x11);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xDD,0x15);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xDE,0x19);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xDF,0x10);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xE0,0x17);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xE1,0x0A);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x30);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x52);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x03);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x00,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x01,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x02,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x03,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x04,0x61);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x05,0x80);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x06,0xc7);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x07,0x01);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x08,0x82);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x09,0x83);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x30,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x31,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x32,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x33,0x2A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x34,0x61);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x35,0xc5);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x36,0x80);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x37,0x23);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x40,0x82);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x41,0x83);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x42,0x80);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x43,0x81);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x44,0x11);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x45,0xe6);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x46,0xe5);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x47,0x11);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x48,0xe8);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x49,0xe7);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x50,0x02);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x51,0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x52,0x04);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x53,0x03);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x54,0x11);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x55,0xea);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x56,0xe9);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x57,0x11);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x58,0xec);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x59,0xeb);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x7e,0x02);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x7f,0x80);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xe0,0x5a);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB1,0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB4,0x0e);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB5,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB6,0x04);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB7,0x07);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB8,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xB9,0x05);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xBA,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xC7,0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xCA,0x0e);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xCB,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xCC,0x04);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xCD,0x07);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xCE,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xCF,0x05);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xD0,0x0f);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x81,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x84,0x0e);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x85,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x86,0x07);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x87,0x04);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x88,0x05);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x89,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x8A,0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x97,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x9A,0x0e);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x9B,0x0f);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x9C,0x07);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x9D,0x04);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x9E,0x05);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x9F,0x06);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xA0,0x00);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x30);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x52);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x02);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x01,0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x02,0xDA);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x03,0xBA);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x04,0xA8);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x05,0x9A);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x06,0x70);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x07,0xFF);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x08,0x91);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x09,0x90);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x0A,0xFF);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x0B,0x8F);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x0C,0x60);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x0D,0x58);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x0E,0x48);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x0F,0x38);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x10,0x2B);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x30);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x52);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF,0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x36,0x02);
+
+	return 0;
+}
 static int panel_nv3051d_unprepare(struct drm_panel *panel)
 {
 	struct panel_nv3051d *ctx = panel_to_panelnv3051d(panel);
@@ -269,27 +487,14 @@ static int panel_nv3051d_unprepare(struct drm_panel *panel)
 
 	return 0;
 }
-
-static int panel_nv3051d_prepare(struct drm_panel *panel)
+static int nv3051d_enable(struct drm_panel *panel) 
 {
 	struct panel_nv3051d *ctx = panel_to_panelnv3051d(panel);
 	struct mipi_dsi_device *dsi = to_mipi_dsi_device(ctx->dev);
 	int ret;
 
-	dev_dbg(ctx->dev, "Resetting the panel\n");
-	ret = regulator_enable(ctx->vdd);
-	if (ret < 0) {
-		dev_err(ctx->dev, "Failed to enable vdd supply: %d\n", ret);
-		return ret;
-	}
+	ret = ctx->panel_info->init_sequence(ctx);
 
-	usleep_range(2000, 3000);
-	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
-	msleep(150);
-	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
-	msleep(20);
-
-	ret = panel_nv3051d_init_sequence(ctx);
 	if (ret < 0) {
 		dev_err(ctx->dev, "Panel init sequence failed: %d\n", ret);
 		goto disable_vdd;
@@ -308,7 +513,6 @@ static int panel_nv3051d_prepare(struct drm_panel *panel)
 		dev_err(ctx->dev, "Failed to set display on: %d\n", ret);
 		goto disable_vdd;
 	}
-
 	usleep_range(10000, 15000);
 
 	return 0;
@@ -316,6 +520,26 @@ static int panel_nv3051d_prepare(struct drm_panel *panel)
 disable_vdd:
 	regulator_disable(ctx->vdd);
 	return ret;
+};
+static int panel_nv3051d_prepare(struct drm_panel *panel)
+{
+	struct panel_nv3051d *ctx = panel_to_panelnv3051d(panel);
+	int ret;
+
+	dev_dbg(ctx->dev, "Resetting the panel\n");
+	ret = regulator_enable(ctx->vdd);
+	if (ret < 0) {
+		dev_err(ctx->dev, "Failed to enable vdd supply: %d\n", ret);
+		return ret;
+	}
+
+	usleep_range(2000, 3000);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	msleep(150);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
+	msleep(20);
+
+	return 0;
 }
 
 static int panel_nv3051d_get_modes(struct drm_panel *panel,
@@ -350,6 +574,7 @@ static int panel_nv3051d_get_modes(struct drm_panel *panel,
 }
 
 static const struct drm_panel_funcs panel_nv3051d_funcs = {
+	.enable	   = nv3051d_enable,
 	.unprepare	= panel_nv3051d_unprepare,
 	.prepare	= panel_nv3051d_prepare,
 	.get_modes	= panel_nv3051d_get_modes,
@@ -371,7 +596,7 @@ static int panel_nv3051d_probe(struct mipi_dsi_device *dsi)
 	if (!ctx->panel_info)
 		return -EINVAL;
 
-	ctx->reset_gpio = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_HIGH);
+	ctx->reset_gpio = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_LOW);
 	if (IS_ERR(ctx->reset_gpio)) {
 		dev_err(dev, "cannot get reset gpio\n");
 		return PTR_ERR(ctx->reset_gpio);
@@ -387,10 +612,11 @@ static int panel_nv3051d_probe(struct mipi_dsi_device *dsi)
 
 	mipi_dsi_set_drvdata(dsi, ctx);
 
-	dsi->lanes = 4;
+	dsi->lanes = ctx->panel_info->lanes;
 	dsi->format = MIPI_DSI_FMT_RGB888;
 	dsi->mode_flags = ctx->panel_info->mode_flags;
 
+
 	drm_panel_init(&ctx->panel, &dsi->dev, &panel_nv3051d_funcs,
 		       DRM_MODE_CONNECTOR_DSI);
 
@@ -492,6 +718,21 @@ static const struct drm_display_mode nv3051d_rk2023_modes[] = {
 	},
 };
 
+static const struct drm_display_mode nv3051d_ee0350et_modes[] = {
+	{
+		.hdisplay    = 640,
+		.hsync_start = 640 + 20,
+		.hsync_end   = 640 + 20 + 2,
+		.htotal	     = 640 + 20 + 2 + 20,
+		.vdisplay    = 480,
+		.vsync_start = 480 + 4,
+		.vsync_end   = 480 + 4 + 2,
+		.vtotal	     = 480 + 4 + 2 + 12,
+		.clock	     = 20378,
+		.flags	     = DRM_MODE_FLAG_NHSYNC | DRM_MODE_FLAG_NVSYNC,
+	},
+};
+
 static const struct nv3051d_panel_info nv3051d_rg351v_info = {
 	.display_modes = nv3051d_rgxx3_modes,
 	.num_modes = ARRAY_SIZE(nv3051d_rgxx3_modes),
@@ -501,6 +742,8 @@ static const struct nv3051d_panel_info nv3051d_rg351v_info = {
 	.mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
 		      MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET |
 		      MIPI_DSI_CLOCK_NON_CONTINUOUS,
+	.lanes = 4,
+	.init_sequence = panel_nv3051d_init_sequence,
 };
 
 static const struct nv3051d_panel_info nv3051d_rg353p_info = {
@@ -508,9 +751,11 @@ static const struct nv3051d_panel_info nv3051d_rg353p_info = {
 	.num_modes = ARRAY_SIZE(nv3051d_rgxx3_modes),
 	.width_mm = 70,
 	.height_mm = 57,
+	.lanes = 4,
 	.bus_flags = DRM_BUS_FLAG_DE_LOW | DRM_BUS_FLAG_PIXDATA_DRIVE_NEGEDGE,
 	.mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
 		      MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.init_sequence = panel_nv3051d_init_sequence,
 };
 
 static const struct nv3051d_panel_info nv3051d_rk2023_info = {
@@ -518,15 +763,30 @@ static const struct nv3051d_panel_info nv3051d_rk2023_info = {
 	.num_modes = ARRAY_SIZE(nv3051d_rk2023_modes),
 	.width_mm = 70,
 	.height_mm = 57,
+	.lanes = 4,
 	.bus_flags = DRM_BUS_FLAG_DE_LOW | DRM_BUS_FLAG_PIXDATA_DRIVE_NEGEDGE,
 	.mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
 		      MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.init_sequence = panel_nv3051d_init_sequence,
+};
+
+static const struct nv3051d_panel_info nv3051d_ee0350et_info = {
+	.display_modes = nv3051d_ee0350et_modes,
+	.num_modes = ARRAY_SIZE(nv3051d_ee0350et_modes),
+	.width_mm    = 70,
+	.height_mm   = 52,
+	.lanes = 2,
+	.bus_flags = DRM_BUS_FLAG_DE_LOW | DRM_BUS_FLAG_PIXDATA_DRIVE_NEGEDGE,
+	.mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
+		      MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_VSYNC_FLUSH,
+	.init_sequence = panel_nv3051d_ee0350et_init_sequence,
 };
 
 static const struct of_device_id newvision_nv3051d_of_match[] = {
 	{ .compatible = "anbernic,rg351v-panel", .data = &nv3051d_rg351v_info },
 	{ .compatible = "anbernic,rg353p-panel", .data = &nv3051d_rg353p_info },
 	{ .compatible = "powkiddy,rk2023-panel", .data = &nv3051d_rk2023_info },
+	{ .compatible = "newvision,nv3051d-ee0350et", .data = &nv3051d_ee0350et_info},
 	{ /* sentinel */ }
 };
 MODULE_DEVICE_TABLE(of, newvision_nv3051d_of_match);
-- 
2.40.1

